### PR TITLE
Support different toolchains for different Rust projects

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Currently, it's not possible to have e.g. `error-stack` using a different toolchain than `hEngine`, which leads to problems, especially if `error-stack` should run with multiple toolchains like i.e. `1.63`, `stable`, and `nightly-2022-08-08`

## 🔗 Related links

- #930 

## 🔍 What does this change?

- Change the logic in `setup.py` to generate the matrix for a job, not only matrix elements
- Require `rust-toolchain.toml` for each project (added for `error-stack`)

## 📜 Does this require a change to the docs?

No, setting the MSRV for `error-stack` will be a follow-up PR

### CODE REVIEW

File: .github/scripts/rust/setup.py [hunk 1]
```diff
@@ -101,40 +96,55 @@ def filter_for_publishable_crates(crates):
...
     :param crates: a list of paths to crates
     """
 
-    output = json.dumps([dict(toolchain="stable", directory=str(crate)) for crate in filter_for_nightly_only_crates(crates)])
-    print(f"::set-output name=exclude::{output}")
-    print(f"exclude = {output}")
+    available_toolchains = []
+    for crate in crates:
+        for pattern, additional_toolchains in TOOLCHAINS.items():
+            toolchain_toml = open(crate/"rust-toolchain.toml", "r")
```
Comment (thehabbos007) [inline .github/scripts/rust/setup.py:109]: Curious about this syntax. Does pathlib overload the `/` operator to be path concat?

Comment (TimDiekmann) [reply .github/scripts/rust/setup.py:109]: Yes 🙂

**Review by thehabbos007 — COMMENTED** [review]
Looks good to me, but I haven't dealt with the rust part of CI a lot, so I'll leave this as a comment

File: .github/scripts/rust/setup.py [hunk 2]
```diff
@@ -101,40 +96,55 @@ def filter_for_publishable_crates(crates):
...
 def main():
     diffs = generate_diffs()
     available_crates = find_local_crates()
     changed_crates = list(set(filter_for_changed_crates(diffs, available_crates)))
 
-    output("lint", changed_crates)
-    output("test", changed_crates)
-    output("publish", filter_crates_by_changed_version(diffs, filter_for_publishable_crates(changed_crates)))
-    output_exclude(changed_crates)
+    output_matrix("lint", changed_crates)
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply .github/scripts/rust/setup.py:145]: Yes, I removed the `development` profile for lints (it was unused and also shouldn't affect production builds anyway). Same applies to `production` on the publish job. It was only used for testing which was already covered in the tests job.

File: .github/scripts/rust/setup.py (hunk 1 above)
Comment (Alfred-Mountfield) [reply .github/scripts/rust/setup.py:109]: It does and I love it. I think generally people put a space around the operator though to be clearer

Comment (TimDiekmann) [reply .github/scripts/rust/setup.py:109]: Is this a change request? ^ 🙃

Comment (Alfred-Mountfield) [reply .github/scripts/rust/setup.py:109]: I was going to do a review tomorrow morning, but yes I think I am asking for a space 😁 

We should consider running pylint on this locally or something

Comment (TimDiekmann) [reply .github/scripts/rust/setup.py]: Applied pylons and reformatted the code: fa2964e36acc8b7957ec9c005238b05de96a20d8

File: .github/scripts/rust/setup.py [hunk 3]
```diff
@@ -2,14 +2,18 @@
...
+import itertools
 
 CWD = Path.cwd()
 
 # All jobs for all crates will run if any of these paths change
 ALWAYS_RUN_PATTERNS = ["**/rust-toolchain.toml", ".github/**"]
 
-# Exclude the stable channel for these crates
-DISABLE_STABLE_PATTERNS = ["packages/engine**", "packages/graph/hash_graph**"]
+# Toolchains used for the specified crates in addition to the default toolchain (rust-toolchain.toml)
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply .github/scripts/rust/setup.py]: Each project can have an independent `rust-toolchain.toml`, so e.g. hEngine could have `channel = "nightly-2022-01-01"` and `error-stack` could use `channel = nightly-2022-08-08`. In addition, you can specify toolchains which should run for a crate, so i.e. `error-stack` will run for `1.61` *and* `nightly-2022-08-08`

[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply .github/scripts/rust/setup.py]: So, just to triple check I understand. We _can_ now have a different toolchain version for the engine and for `error-stack`. Which before we couldn't? Correct?

[the remark this answers is not in the document]
Comment (TimDiekmann) [reply .github/scripts/rust/setup.py]: Yes 🙂

[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply .github/scripts/rust/setup.py]: Awesome

File: .github/scripts/rust/setup.py (hunk 1 above)
Comment (Alfred-Mountfield) [reply .github/scripts/rust/setup.py]: Thanks

**Review by Alfred-Mountfield — APPROVED** [review]
Cheers Tim! This was bothering me :D

**Pull request merged**